qualtran/bloqs/arithmetic/controlled_add_or_subtract.py

```diff
@@ -0,0 +1,125 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import TYPE_CHECKING, Union
+
+from attrs import field, frozen
+
+from qualtran import Bloq, bloq_example, BloqDocSpec, QBit, QInt, QMontgomeryUInt, QUInt, Signature
+from qualtran.bloqs.arithmetic.addition import Add
+from qualtran.bloqs.arithmetic.bitwise import BitwiseNot
+from qualtran.bloqs.basic_gates import XGate
+
+if TYPE_CHECKING:
+    from qualtran import BloqBuilder, Soquet, SoquetT
+
+
+@frozen
+class ControlledAddOrSubtract(Bloq):
+    r"""Adds or subtracts in-place into the target, based on a control bit.
+
+    Applies the transformation
+
+    $$
+        |1\rangle |a\rangle |b\rangle \mapsto |1\rangle |a\rangle |b + a\rangle \\
+        |0\rangle |a\rangle |b\rangle \mapsto |0\rangle |a\rangle |b - a\rangle
+    $$
+
+    Given two numbers `a`, `b` and a control bit `ctrl`, this bloq computes:
+    - the sum `b + a` when `ctrl=1`,
+    - the difference `b - a` when `ctrl=0`,
+    and stores it in the second register (`b`).
```

Comment:
markdown nit: this line gets rendered as part of the second bullet point. You'd need a newline to separate it

Reply:
oops, thanks! added a newline